stan/math/prim/mat/fun/log_mix.hpp

```diff
@@ -94,6 +97,274 @@ typename return_type<T_theta, T_lam>::type log_mix(const T_theta& theta,
 
   return ops_partials.build(logp);
 }
+
+/**
+ * Return the log mixture density given specified mixing proportions
+ * and array of log density vectors.
+ *
+ * \f[
+ * \frac{\partial }{\partial p_x}\left[
+ * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+d_n}\right)+
+ * \log\left(\exp^{\log\left(p_1\right)+f_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+f_n}\right)\right]
+ * =\frac{e^{d_x}}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}+
+ * \frac{e^{f_x}}{e^{f_1}p_1+\cdot\cdot\cdot+e^{f_m}p_m}
+ * \f]
+ *
+ * \f[
+ * \frac{\partial }{\partial d_x}\left[
+ * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+d_n}\right)
+ * +\log\left(\exp^{\log\left(p_1\right)+f_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+f_n}\right)\right]
+ * =\frac{e^{d_x}p_x}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}
+ * \f]
+ *
+ * @param theta vector of mixing proportions in [0, 1].
+ * @param lambda array containing vectors of log densities.
+ * @return log mixture of densities in specified proportion
+ */
+template <typename T_theta, typename T_lam>
+typename return_type<T_theta, std::vector<Eigen::Matrix<T_lam, -1, 1> > >::type
+log_mix(const T_theta& theta,
+        const std::vector<Eigen::Matrix<T_lam, -1, 1> >& lambda) {
+  static const char* function = "log_mix";
+  typedef typename stan::partials_return_type<
+      T_theta, std::vector<Eigen::Matrix<T_lam, -1, 1> > >::type
+      T_partials_return;
+
+  typedef typename Eigen::Matrix<T_partials_return, -1, 1> T_partials_vec;
+
+  typedef typename Eigen::Matrix<T_partials_return, -1, -1> T_partials_mat;
+
+  typedef typename std::vector<Eigen::Matrix<T_lam, -1, 1> > T_lamvec_type;
+
+  const int N = length(lambda);
+  const int M = theta.size();
+
+  check_bounded(function, "theta", theta, 0, 1);
+  check_not_nan(function, "theta", theta);
+  check_finite(function, "theta", theta);
+  for (int n = 0; n < N; ++n) {
+    check_not_nan(function, "lambda", lambda[n]);
+    check_finite(function, "lambda", lambda[n]);
+    check_consistent_sizes(function, "theta", theta, "lambda", lambda[n]);
+  }
+
+  scalar_seq_view<T_theta> theta_vec(theta);
+  T_partials_vec theta_dbl(M);
+  for (int m = 0; m < M; ++m)
+    theta_dbl[m] = value_of(theta_vec[m]);
+
+  T_partials_mat lam_dbl(M, N);
+  vector_seq_view<T_lamvec_type> lam_vec(lambda);
+  for (int n = 0; n < N; ++n)
+    for (int m = 0; m < M; ++m)
+      lam_dbl(m, n) = value_of(lam_vec[n][m]);
+
+  T_partials_mat logp_tmp = log(theta_dbl).replicate(1, N) + lam_dbl;
+
+  T_partials_vec logp(N);
+  for (int n = 0; n < N; ++n)
+    logp[n] = log_sum_exp(logp_tmp.col(n).eval());
+
+  T_partials_mat derivs
+      = (lam_dbl - logp.transpose().replicate(M, 1))
+            .unaryExpr([](T_partials_return x) { return exp(x); });
```

Comment:
This seems less efficient than doing the exp first and then replicating, no?

Reply:
The replicate is only applied to the vector ```logp``` so that ```lam_dbl``` and ```logp``` are matrices of identical dimensions and the subtraction can be done in a single step (essentially a row-wise subtraction of a vector from a matrix). The exp operation is then applied to each element of the resulting matrix.
i.e.: ```exp(lam_dbl - logp.transpose().replicate(M, 1))```

Reply:
Right - isn't this more efficient:
```
exp(lam_dbl - logp.transpose()).replicate(M, 1)
```
(notice the placement of the parens)

Reply:
Hmm, seems like I'm wrong about that. I guess replicate does some special sauce under the hood. The weirdest thing is that my benchmark actually thinks the first version (exp'ing the replicate instead of replicating the exp'd) [is way faster](https://github.com/seantalts/cppperf/blob/master/eigen_replicate_results.txt)

Reply:
Wait, no, I got mixed up in the results. "ReplicateAfter" is way faster, i.e. we should call exp first, i.e. [the 2nd version](https://github.com/stan-dev/math/pull/751#discussion_r176154760) with the paren moved inwards is better.

Reply:
But that's not a valid operation? ```lam_dbl``` is a matrix with dimensions [M,N] and ```logp``` is a vector with dimensions [N,1] (making ```logp.transpose()``` a row vector with dimensions [1,N].

```([M,N] - [1,N])``` is not valid arithmetic in eigen (https://eigen.tuxfamily.org/dox/group__TutorialMatrixArithmetic.html): 
>Addition and subtraction
>The left hand side and right hand side must, of course, have the same numbers of rows and of columns

The ```replicate()``` is needed so that ```logp.transpose()``` has dimensions [M,N] and the resulting operation is ```([M,N] - [M,N])```

Attempting to evaluate ```exp(lam_dbl - logp.transpose()).replicate(M, 1) ``` returns:
```
Assertion `(aLhs.rows() == aRhs.rows() && aLhs.cols() == aRhs.cols())` failed.
Aborted (core dumped)
```

Reply:
Oh, you're right, didn't notice the difference in dimensionality. Sorry for the confusion.